Modified some things to make use more consistent with traditional editors

jasper-lu opened the pull request:
Hi, I used your library for a docx editor app I was working on. Thanks for the sweet library!

I made a few changes which I think would improve your editor while I was using it, and I wanted to see if you would like to add them to your own editor as well.

I changed the boolean effects and style effects flags from SPAN_EXCLUSIVE_EXCLUSIVE to SPAN_EXCLUSIVE_INCLUSIVE instead, making it more similar to regular editors:

![](http://i.imgur.com/4sSBK7T.png)![](http://i.imgur.com/OfwGoBA.png)

![](http://i.imgur.com/Ttxjleb.png)

The main change is that appended text keeps the style.

I also changed the flags for Alignment from SPAN_INCLUSIVE_INCLUSIVE to SPAN_PARAGRAPH. This makes it work a bit bitter / more logical in the way it works:

![](http://i.imgur.com/zgN6e7v.png)
![](http://i.imgur.com/OSlXzgb.png)

To get those results, I typed:

This is the

old *right aligned old*

way *hit left align on this line*

In the old way, the last left align would affect both this line and the preceding line. In the new way, the right align still sticks when going to new lines, but hitting left align affects only the line it was pressed on.

commonsguy commented:
I can run some tests and consider changing those flags. Your pull request changes other things though.

jasper-lu commented:
I changed a few things to make it work in an EXCLUSIVE_INCLUSIVE format.

I.e, if I'm working on a bolded text, and then hit bold again, the span is changed back to EXCLUSIVE_EXCLUSIVE, so that you could type normally again...

However, I reverted the SPAN_PARAGRAPH code. I didn't test it enough, and it didn't work when I tried plugging that into my app in lieu of my original code.

I was working with an old version of your library, and had to modify line alignment heavily to get it to work the way I wanted.

commonsguy commented:
> I was working with an old version of your library, and had to modify line alignment heavily to get it to work the way I wanted.

Yeah, the line alignment stuff was pretty much just a stab in the dark prior to this weekend's 0.5.0. It's still in need of work, but I wanted to get an update out.

Anyway, I will ponder the `SPAN_EXCLUSIVE_EXCLUSIVE` change. Thanks!

jakubkrolewski commented:
Unfortunately, it doesn't always work well. As for me, it looks like a bug in Android spans. The problem is as follows:
1. Type "abc"
2. Select "c"
3. Apply bold style to it (you can use the demo activity that I created in my fork - https://github.com/jakubkrolewski/cwac-richedit/tree/feature/operations_on_not_selected_text - it has this pull  request already merged)
4. Move cursor to the end (after "c")
5. Type "d"

In theory, the letters "bc" should now be bolden, but I my on Nexus 5 (Android 4.4.4) none of them are. After debugging the code I found out that the style span length was changed to zero and it was moved to the end of the string.

pchupik commented:
@jakubkrolewski, that's probably because auto correct feature of your keyboard.
Try disabling it

jakubkrolewski commented:
I turned the auto correction off, but it didn't change anything.

jasper-lu commented:
That's because this is using SPAN_EXCLUSIVE_INCLUSIVE. iirc, this makes
everything before the span say the same, but everything after have the
style ( if you bold something on Google docs, it should work in a similar
way).

On Tue, Apr 28, 2015, 8:02 AM jakubkrolewski <notifications@github.com>
wrote:

> I turned the auto correction off, but it didn't change anything.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/commonsguy/cwac-richedit/pull/14#issuecomment-97035690>
> .
>

jakubkrolewski commented:
@jasper-lu Yes, it should work as you described it, but the problem is that it doesn't. In my example there was a little error, which I've just corrected. Please, revise it once again.
After typing a new letter ("d") after the one that is bolden ("c"), the new letter should become bolden as well. In fact, it doesn't and what's more, the bold style disappears from the first letter ("c").

pchupik commented:
@jakubkrolewski I am pretty sure that's because of auto correction 
(Which is not so easy to turn off completely)
Adding 

        android:inputType="textMultiLine|textVisiblePassword|textNoSuggestions"

to layout of RichEditText view solves the issue.
Try it.

jakubkrolewski commented:
@pchupik You're right, that was it, thanks :)
I think that this PR should be updated with setting this input type in code.
Apart from that, I noticed that the spans flags were not modified in the RelativeSizeEffect.

jasper-lu commented:
@jakubkrolewski ah my fault.

Also, this was intended to be more if a guideline at first, since I didn't
know if cwac wanted this style of text editing.

I chose to use these because they are boolean effects for the most part (I
made it so that pressing the effect again when on a bolded phrase makes the
span exclusive_exclusive instead.

On Wed, Apr 29, 2015, 9:18 AM jakubkrolewski <notifications@github.com>
wrote:

> @pchupik <https://github.com/pchupik> You're right, that was it, thanks :)
> I think that this PR should be updated with setting this input type in
> code.
> Apart from that, I noticed that the spans flags were not modified in the
> RelativeSizeEffect.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/commonsguy/cwac-richedit/pull/14#issuecomment-97422841>
> .
>

jakubkrolewski commented:
I think now that adding the textNoSuggestions is not a good solution. Having suggestions is extremely useful when typing long texts and we should not sacrifice this feature.
If the spans flags don't work without it, another way should be found.
In this project https://github.com/jakubkrolewski/droid-writer all spans are exclusive, but they are manually adjusted after typing each character. Maybe this solution will work better.

jakubkrolewski commented:
After a lot of experiments with different approaches, I think there's no easy way to fix the problems I described earlier. They occur only if a character style is changed in the middle of a word, so standard users should never encounter it.
All in all, I think that this PR should be merged. The solution is not perfect, but it's a big step forward.
